### PR TITLE
[DI-75][Indexer] Create Token V2 data models and convert token v1

### PR DESCRIPTION
**Pull request by bowenyang007**
### Description
Create data models according to https://dbdiagram.io/d/6433a3808615191cfa8c974e (I'll create a proper wiki documenting every field)
High level, we have these tables
* objects / current_objects
* collections_v2, current_collections_v2
* token_datas_v2, current_token_datas_v2
* token_ownerships_v2, current_token_ownerships_v2

The new tables include both **token v1 and token v2**. Some files are token v1 or token v2 exclusive (those fields should be obvious since they are suffixed with `_v1` or `v2`).

### What's missing?
* V2 activity not done yet
* Fungible token not done. Waiting for contract to finalize. 

### Test Plan
Will provide hasura links, but here are some screenshots

image

image

confirmed that token properties work 
image

### CODE REVIEW

File: crates/indexer/src/models/move_resources.rs [hunk 1]
```diff
@@ -23,6 +23,7 @@ pub struct MoveResource {
     pub generic_type_params: Option<serde_json::Value>,
     pub data: Option<serde_json::Value>,
     pub is_deleted: bool,
+    pub state_key_hash: String,
```
Comment (movekevin) [inline crates/indexer/src/models/move_resources.rs:26]: What is this for again?

File: crates/indexer/src/models/v2_objects.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    pub is_deleted: bool,
+}
+
+#[derive(Debug, Deserialize, FieldCount, Identifiable, Insertable, Serialize)]
+#[diesel(primary_key(object_address))]
+#[diesel(table_name = current_objects)]
+pub struct CurrentObject {
+    pub object_address: String,
+    pub owner_address: Option<String>,
+    pub state_key_hash: String,
```
Comment (movekevin) [inline crates/indexer/src/models/v2_objects.rs:41]: What is this used for?

File: crates/indexer/src/models/v2_objects.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            write_resource.data.typ.address,
+            write_resource.data.typ.module,
+            write_resource.data.typ.name
+        );
+        if !V2TokenResource::is_resource_supported(type_str.as_str()) {
+            return Ok(None);
+        }
+        let resource = MoveResource::from_write_resource(
+            write_resource,
+            0, // Placeholder, this isn't used anyway
```
Comment (movekevin) [inline crates/indexer/src/models/v2_objects.rs:96]: What are these?

File: crates/indexer/src/models/token_models/v2_token_utils.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum V2TokenResource {
+    ObjectCore(ObjectCore),
+}
+
+impl V2TokenResource {
+    pub fn is_resource_supported(data_type: &str) -> bool {
+        matches!(data_type, "0x1::object::ObjectCore")
```
Comment (movekevin) [inline crates/indexer/src/models/token_models/v2_token_utils.rs:27]: This is not right. This should be looking at 0x4::token_objects::Token as well. 0x1::object::ObjectCore is for all objects

File: crates/indexer/src/models/move_resources.rs (hunk 1 above)
Comment (gregnazario) [reply crates/indexer/src/models/move_resources.rs:26]: This is the differentiator between resources in a resource group.

File: crates/indexer/src/processors/default_processor.rs [hunk 5]
```diff
@@ -460,15 +522,29 @@ impl TransactionProcessor for DefaultTransactionProcessor {
...
+            all_current_objects.extend(current_objects);
+        }
         // Getting list of values and sorting by pk in order to avoid postgres deadlock since we're doing multi threaded db writes
         let mut current_table_items = current_table_items
             .into_values()
             .collect::<Vec<CurrentTableItem>>();
         let mut table_metadata = table_metadata.into_values().collect::<Vec<TableMetadata>>();
+        let mut all_current_objects = all_current_objects
+            .into_values()
+            .collect::<Vec<CurrentObject>>();
```
Comment (gregnazario) [inline crates/indexer/src/processors/default_processor.rs:541]: I'm curious, why do you make a hashmap of objects, and then drop the keys?  Do you get duplicates?

File: crates/indexer/src/models/v2_objects.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                    is_deleted: true,
+                },
+                CurrentObject {
+                    object_address: resource.address,
+                    owner_address: None,
+                    state_key_hash: resource.state_key_hash,
+                    allow_ungated_transfer: None,
+                    last_guid_creation_num: None,
+                    last_transaction_version: txn_version,
+                    is_deleted: false,
```
Comment (gregnazario) [inline crates/indexer/src/models/v2_objects.rs:160]: I wonder how are deletes handled from the object side?

File: crates/indexer/src/models/v2_objects.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    pub last_transaction_version: i64,
+    pub is_deleted: bool,
+}
+
+impl Object {
+    /// Only parsing 0x1 ObjectCore from transactions
+    pub fn from_transaction(
+        transaction: &Transaction,
+    ) -> (Vec<Self>, HashMap<String, CurrentObject>) {
+        if let Transaction::UserTransaction(user_txn) = transaction {
```
Comment (gregnazario) [inline crates/indexer/src/models/v2_objects.rs:53]: FYI, today we don't handle GenesisTransactions, but you may need to handle them in the future.

A GenesisTransaction is a writeset transaction... and isn't always at genesis

File: crates/indexer/src/models/move_resources.rs (hunk 1 above)
Comment (bowenyang007) [reply crates/indexer/src/models/move_resources.rs:26]: yea we need this to know what's in a resource group. It's the identifier for slot I think.

File: crates/indexer/src/models/v2_objects.rs (hunk 3 above)
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:96]: nothing. for logging.

File: crates/indexer/src/models/token_models/v2_token_utils.rs (hunk 4 above)
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_token_utils.rs:27]: This is just for objects. we'll handle tokens separately.

File: crates/indexer/src/processors/default_processor.rs (hunk 5 above)
Comment (bowenyang007) [reply crates/indexer/src/processors/default_processor.rs:541]: This is current objects across a list of transactions.

File: crates/indexer/src/models/v2_objects.rs (hunk 6 above)
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:160]: ugh.

File: crates/indexer/src/models/v2_objects.rs (hunk 7 above)
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:53]: objects aren't in genesis right? I do handle genesis transactions just case by case.

File: crates/indexer/src/models/move_resources.rs (hunk 1 above)
Comment (movekevin) [reply crates/indexer/src/models/move_resources.rs:26]: How is this used?

File: crates/indexer/src/models/stake_models/delegator_pools.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+// All pools
+#[derive(Debug, Deserialize, FieldCount, Identifiable, Insertable, Serialize)]
+#[diesel(primary_key(staking_pool_address))]
+#[diesel(table_name = delegated_staking_pools)]
+pub struct DelegatorPool {
+    pub staking_pool_address: String,
+    pub first_transaction_version: i64,
+}
+
+impl DelegatorPool {
```
Comment (movekevin) [inline crates/indexer/src/models/stake_models/delegator_pools.rs:26]: This looks like it belongs in a separate PR?

File: crates/indexer/src/models/token_models/v2_token_utils.rs (hunk 4 above)
Comment (movekevin) [reply crates/indexer/src/models/token_models/v2_token_utils.rs:27]: But this is a function on V2TokenResource? Shouldn't this be in something that implements Object?

File: crates/indexer/src/models/v2_objects.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+use aptos_api_types::{DeleteResource, Transaction, WriteResource, WriteSetChange};
+use bigdecimal::BigDecimal;
+use field_count::FieldCount;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+#[derive(Debug, Deserialize, FieldCount, Identifiable, Insertable, Serialize)]
+#[diesel(primary_key(transaction_version, write_set_change_index))]
+#[diesel(table_name = objects)]
+pub struct Object {
```
Comment (movekevin) [inline crates/indexer/src/models/v2_objects.rs:23]: Should we add a field that lists the name of all resources that are on each object like we discussed yesterday? These can be used to "identify" an object and filter for specific objects that an account currently owns for example

File: crates/indexer/src/models/v2_objects.rs (hunk 7 above)
Comment (movekevin) [reply crates/indexer/src/models/v2_objects.rs:52]: Yeah this is likely lower in priority as we don't use GenesisTxs for some sort of checkpointing the network yet. And the first Genesis tx wouldn't have objects.

File: crates/indexer/src/models/move_resources.rs (hunk 1 above)
Comment (bowenyang007) [reply crates/indexer/src/models/move_resources.rs:26]: Not sure rn. Just creating the data models. Probably will need to look up to know what items are deleted.

File: crates/indexer/src/models/v2_objects.rs (hunk 9 above)
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:23]: Let's discuss how to best do this moving. I'm leaning towards using Cassandra. I don't think that postgres is the right db choice for this.

Comment (movekevin) [inline crates/indexer/src/models/v2_objects.rs:23]: I thought we also want to track the object's "types" - i.e. all the resources defined in the ObjectCore resource group?

File: crates/indexer/src/schema.rs [hunk 10]
```diff
@@ -201,6 +255,25 @@ diesel::table! {
...
+        collection_id -> Varchar,
+        token_name -> Varchar,
+        maximum -> Numeric,
+        supply -> Numeric,
+        largest_property_version_v1 -> Nullable<Numeric>,
+        token_uri -> Varchar,
+        description -> Text,
+        token_properties -> Jsonb,
+        token_standard -> Varchar,
+        is_fungible_v2 -> Nullable<Bool>,
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/schema.rs:270]: is_fungible is a v2 thing only. V1 has no fungibility flag.

File: crates/indexer/src/processors/token_processor.rs [hunk 11]
```diff
@@ -520,15 +785,13 @@ impl TransactionProcessor for TokenTransactionProcessor {
         let table_handle_to_owner =
             TableMetadataForToken::get_table_handle_to_owner_from_transactions(&transactions);
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/processors/token_processor.rs:762]: no tables in v2 so we don't need this.

File: crates/indexer/src/models/v2_objects.rs [hunk 12]
```diff
@@ -0,0 +1,166 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+// This is required because a diesel macro makes clippy sad
+#![allow(clippy::extra_unused_lifetimes)]
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:5]: this is to help me find the files quickly lol. i might rename when we migrate to grpc.

File: crates/indexer/src/models/v2_objects.rs (hunk 9 above)
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:26]: Maybe, but I prolly need to do it somewhere else not here.

File: crates/indexer/src/schema.rs [hunk 13]
```diff
@@ -427,6 +532,28 @@ diesel::table! {
...
+        owner_address -> Varchar,
+        collection_id -> Varchar,
+        token_data_id -> Nullable<Varchar>,
+        property_version_v1 -> Nullable<Numeric>,
+        #[sql_name = "type"]
+        type_ -> Varchar,
+        from_address -> Nullable<Varchar>,
+        to_address -> Nullable<Varchar>,
+        token_amount -> Nullable<Numeric>,
+        before_value -> Nullable<Text>,
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/schema.rs]: this is like if we change the uri from X to Y, what's the old URI vs new URI.

File: crates/indexer/src/schema.rs [hunk 14]
```diff
@@ -472,6 +619,24 @@ diesel::table! {
     }
 }
 
+diesel::table! {
+    token_ownerships_v2 (transaction_version, write_set_change_index) {
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/schema.rs:624]: yes if we want to see historical ownership.

File: crates/indexer/src/schema.rs [hunk 15]
```diff
@@ -96,6 +96,27 @@ diesel::table! {
...
+        collection_name -> Varchar,
+        description -> Text,
+        uri -> Varchar,
+        current_supply -> Numeric,
+        max_supply -> Numeric,
+        total_minted_v2 -> Nullable<Numeric>,
+        mutable_description -> Bool,
+        mutable_uri -> Bool,
+        table_handle_v1 -> Nullable<Varchar>,
+        token_standard -> Varchar,
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/schema.rs:114]: Idk what we want to call token standard. What if it's called Token Next at some point lol.

File: crates/indexer/src/schema.rs [hunk 16]
```diff
@@ -96,6 +96,27 @@ diesel::table! {
...
+        transaction_version -> Int8,
+        write_set_change_index -> Int8,
+        collection_id -> Varchar,
+        creator_address -> Varchar,
+        collection_name -> Varchar,
+        description -> Text,
+        uri -> Varchar,
+        current_supply -> Numeric,
+        max_supply -> Numeric,
+        total_minted_v2 -> Nullable<Numeric>,
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/schema.rs:110]: B/c this doesn't exist in v1. If this is just supply...then isn't this the same as current_supply?

File: crates/indexer/src/models/token_models/token_utils.rs [hunk 17]
```diff
@@ -48,6 +52,10 @@ impl TokenDataIdType {
     pub fn get_collection_data_id_hash(&self) -> String {
         CollectionDataIdType::new(self.creator.clone(), self.collection.clone()).to_hash()
     }
+
+    pub fn get_collection_id(&self) -> String {
```
Comment (rtso) [inline crates/indexer/src/models/token_models/token_utils.rs:56]: Why do we need both `get_collection_id` and `get_collection_data_id_hash`?

Comment (bowenyang007) [reply crates/indexer/src/models/token_models/token_utils.rs:56]: get_collection_id is b/c v2 actually has addresses now so we should have standardized pattern. Still have to keep both for backwards compatibility.

File: crates/indexer/src/models/token_models/v2_token_utils.rs [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    pub owner: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum V2TokenResource {
+    ObjectCore(ObjectCore),
+}
+
+impl V2TokenResource {
+    pub fn is_resource_supported(data_type: &str) -> bool {
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_token_utils.rs:321]: Haven't thought deeply enough about writing this code in a cleaner way tbh. Originally decided to do this to avoid converting it to MoveResources all the time, but this can likely be optimized.

File: crates/indexer/src/models/token_models/v2_collections.rs [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+                    token_standard: TokenStandard::V1.to_string(),
+                    transaction_timestamp: txn_timestamp,
+                },
+                CurrentCollectionV2 {
+                    collection_id,
+                    creator_address: collection_id_struct.creator,
+                    collection_name,
+                    description: collection_data.description,
+                    uri,
+                    current_supply: collection_data.supply,
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_collections.rs:248]: why? linter is fine with it lol.

**Review by larry-aptos — APPROVED** [review]

**Review by larry-aptos — APPROVED** [review]

File: crates/indexer/src/models/token_models/v2_collections.rs [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+        Err(anyhow::anyhow!("Failed to get collection creator"))
+    }
+
+    /// TODO: Change this to a KV store
+    pub fn get_by_table_handle(
+        conn: &mut PgPoolConnection,
+        table_handle: &str,
+    ) -> anyhow::Result<String> {
+        let mut res: Vec<Option<CreatorFromCollectionTableV1>> = sql_query(
+            "SELECT creator_address FROM current_collections_v2 WHERE table_handle_v1 = $1",
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_collections.rs:290]: For token standard v2 yea, but we shouldn't be doing this for v2.

File: crates/indexer/src/models/token_models/v2_token_datas.rs [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+            0, // Placeholder, this isn't used anyway
+            txn_version,
+            0, // Placeholder, this isn't used anyway
+        );
+
+        if let V2TokenResource::Token(inner) =
+            V2TokenResource::from_resource(&type_str, resource.data.as_ref().unwrap(), txn_version)?
+        {
+            // Get maximum, supply, and is fungible from fungible asset if this is a fungible token
+            let (maximum, supply, is_fungible_v2) = (None, BigDecimal::zero(), Some(false));
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_token_datas.rs:91]: When fungible assets get into testnet.

**Review by rtso — APPROVED** [review]

File: crates/indexer/src/models/token_models/v2_token_datas.rs [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+                    supply: supply.clone(),
+                    largest_property_version_v1: None,
+                    token_uri: token_uri.clone(),
+                    token_properties: token_properties.clone(),
+                    description: inner.description.clone(),
+                    token_standard: TokenStandard::V2.to_string(),
+                    is_fungible_v2,
+                    transaction_timestamp: txn_timestamp,
+                },
+                CurrentTokenDataV2 {
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_token_datas.rs:127]: b/c of the way that borrowing works in rust lol. A value can be consumed once. Those values haven't been consumed yet so cloning them would just be a waste of memory. Just run `cargo xclippy` and it'll tell you to remove the clones.

File: crates/indexer/src/models/token_models/v2_token_ownerships.rs [hunk 23]
```diff
@@ -0,0 +1,462 @@
...
+        tokens_burned: &TokenV2Burned,
+    ) -> anyhow::Result<Option<(Self, CurrentTokenOwnershipV2)>> {
+        if let Some(token_address) =
+            tokens_burned.get(&standardize_address(&write_resource.address.to_string()))
+        {
+            if let Some(object_core) = ObjectCore::from_write_resource(write_resource, txn_version)?
+            {
+                let token_data_id = token_address.clone();
+                let owner_address = object_core.owner.clone();
+                let storage_id = token_data_id.clone();
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_token_ownerships.rs:162]: Yes for NFTs. For fungible assets they won't be the same.

File: crates/indexer/src/models/token_models/v2_token_ownerships.rs [hunk 24]
```diff
@@ -0,0 +1,462 @@
...
+            &table_item_data.value,
+            txn_version,
+        )? {
+            Some(TokenWriteSet::Token(inner)) => Some(inner),
+            _ => None,
+        };
+
+        if let Some(token) = maybe_token {
+            let table_handle = standardize_address(&table_item.handle.to_string());
+            let amount = ensure_not_negative(token.amount);
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/v2_token_ownerships.rs:283]: Bad data.

File: crates/indexer/src/models/v2_objects.rs [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+                        Self::from_delete_resource(inner, txn_version, index).unwrap()
+                    },
+                    WriteSetChange::WriteResource(inner) => {
+                        Self::from_write_resource(inner, txn_version, index).unwrap()
+                    },
+                    _ => None,
+                };
+                if let Some((object, current_object)) = maybe_object_combo {
+                    objects.push(object);
+                    current_objects.insert(current_object.object_address.clone(), current_object);
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:73]: No but the same object shouldn't appear twice. Write sets are deduped.

File: crates/indexer/src/models/v2_objects.rs [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+                    is_deleted: true,
+                },
+                CurrentObject {
+                    object_address: resource.address,
+                    owner_address: None,
+                    state_key_hash: resource.state_key_hash,
+                    allow_ungated_transfer: None,
+                    last_guid_creation_num: None,
+                    last_transaction_version: txn_version,
+                    is_deleted: false,
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:153]: Yes. I'm not handling that well right now.

[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/v2_objects.rs:153]: oh this needs to say true lmao. thanks.

File: crates/indexer/src/models/property_map.rs [hunk 27]
```diff
@@ -52,3 +52,53 @@ impl PropertyMap {
...
+}
+
+impl TokenObjectPropertyMap {
+    /// Deserializes PropertyValue from bcs encoded json
+    pub fn from_bcs_encode_str(val: Value) -> Option<Value> {
+        let mut pm = TokenObjectPropertyMap {
+            data: HashMap::new(),
+        };
+        let records: &Vec<Value> = val.get("data")?.as_array()?;
+        for entry in records {
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/property_map.rs:84]: @lightmark wrote this code.

[the remark this answers is not in the document]
Comment (lightmark) [reply crates/indexer/src/models/property_map.rs:84]: there is only one entry which is `data`

[the remark this answers is not in the document]
Comment (larry-aptos) [reply crates/indexer/src/models/property_map.rs:84]: also each entry in `data` must contain `key` and `value.value` and `value.type` right?

File: crates/indexer/src/models/token_models/v2_token_utils.rs [hunk 28]
```diff
@@ -0,0 +1,396 @@
...
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Collection {
+    pub creator: String,
+    pub description: String,
```
[the remark this answers is not in the document]
Comment (rtso) [reply crates/indexer/src/models/token_models/v2_token_utils.rs:91]: If it's free form text, should this also be truncated too then like what we do for `name`?

File: crates/indexer/src/models/property_map.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (lightmark) [reply crates/indexer/src/models/property_map.rs:84]: yes. if anyone doesn't work, it may be broken.
